libs/geo/src/main/java/org/elasticsearch/geometry/simplify/GeometrySimplifier.java

```diff
@@ -0,0 +1,482 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.geometry.simplify;
+
+import org.elasticsearch.geometry.Geometry;
+import org.elasticsearch.geometry.Line;
+import org.elasticsearch.geometry.LinearRing;
+import org.elasticsearch.geometry.MultiPolygon;
+import org.elasticsearch.geometry.Polygon;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.stream.Collectors;
+
+public abstract class GeometrySimplifier<T extends Geometry> {
+    protected final int maxPoints;
+    protected final SimplificationErrorCalculator calculator;
+    protected final PointError[] points;
+    protected PointError lastRemoved;
+    protected final Monitor monitor;
+    protected int length;
+    protected int objCount = 0;
+    protected String description;
+
+    protected final PriorityQueue<PointError> queue = new PriorityQueue<>();
+
+    protected GeometrySimplifier(String description, int maxPoints, SimplificationErrorCalculator calculator, Monitor monitor) {
+        this.description = description;
+        this.maxPoints = maxPoints;
+        this.calculator = calculator;
+        this.monitor = monitor;
+        this.points = new PointError[maxPoints];
+        this.length = 0;
+    }
+
+    /**
+     * Simplify an entire geometry in a non-streaming fashion
+     */
+    public abstract T simplify(T geometry);
```

Comment:
I think this is wrong as we are still using the same streaming algorithm. This is just a utility method to perform the streaming algorithm on a geometry?

Reply:
OK. What I did from the beginning is target two use cases, streaming and non-streaming. I also have a bunch of tests asserting that only very limited geometry types can access the streaming case (simple LineString, Polygon with no holes), while all other geometry types can use the non-streaming API. This was because I perceived the non-streaming use cases as very valuable, even if not the first use case to be implemented.

Also, I use the non-streaming API for most tests, and benchmarks as well as the visualisation tool (images and animations used in the presentation I gave).